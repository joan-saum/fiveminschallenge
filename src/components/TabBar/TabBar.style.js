import {StyleSheet} from 'react-native';

const style = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-around',
  },
});

export default style;
