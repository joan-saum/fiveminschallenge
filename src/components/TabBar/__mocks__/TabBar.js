// @flow

import React from 'react';
import {View} from 'react-native';

import type {Node} from 'react';

const MyMockComponent = (): Node => <View />;

export default MyMockComponent;
