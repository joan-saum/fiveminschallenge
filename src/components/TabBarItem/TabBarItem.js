// @flow

import React, {useCallback} from 'react';
import {TouchableOpacity, View, Text} from 'react-native';

import type {Node} from 'react';

import styles from './TabBarItem.style';

type Props = {|
  title: string,
  onPress: (title: string) => void,
|};

const TabBarItem = ({title, onPress}: Props): Node => {
  const onNavPress = useCallback(() => {
    onPress(title);
  }, [onPress, title]);

  return (
    <TouchableOpacity onPress={onNavPress}>
      <View style={styles.container}>
        <Text style={styles.text}>{title}</Text>
      </View>
    </TouchableOpacity>
  );
};

export default TabBarItem;
