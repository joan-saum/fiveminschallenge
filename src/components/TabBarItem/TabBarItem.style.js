import {StyleSheet} from 'react-native';

const style = StyleSheet.create({
  container: {
    height: 64,
    justifyContent: 'center',
  },
  text: {
    color: 'white',
  },
});

export default style;
