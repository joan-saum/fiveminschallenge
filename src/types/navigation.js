// @flow

export type NavigationProps = {|
  setParams: Function,
  getParam: Function,
  navigate: Function,
  dangerouslyGetParent: Function,
  goBack: Function,
  addListener: Function,
  isFocused: Function,
  push: Function,
  pop: Function,
  popToTop: Function,
  dismiss: Function,
  state: Object,
|};
