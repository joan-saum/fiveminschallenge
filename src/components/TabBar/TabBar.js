// @flow

import React, {useCallback} from 'react';
import {View} from 'react-native';
import type {Node} from 'react';

import type {NavigationProps} from '../../types/navigation';

import TabBarItem from '../TabBarItem/TabBarItem';

import styles from './TabBar.style';

type Props = {|
  navigation: NavigationProps,
  routeNames: $ReadOnlyArray<string>,
|};

const TabBar = ({navigation, routeNames}: Props): Node => {
  const onPress = useCallback(
    (route: string) => {
      navigation.navigate(route);
    },
    [navigation],
  );

  return (
    <View style={styles.container}>
      {routeNames.map(r => (
        <TabBarItem key={r} title={r} onPress={onPress} />
      ))}
    </View>
  );
};

export default TabBar;
