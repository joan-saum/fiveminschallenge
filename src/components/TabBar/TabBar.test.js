// @flow

import * as React from 'react';
import renderer from 'react-test-renderer';

import navigation from '../../utils/mockNavigation';
import TabBarItem from '../TabBarItem/TabBarItem';

import TabBar from './TabBar';

jest.unmock('./TabBar');

describe('TabBar', () => {
  it('should render a TabBar', () => {
    const routeNames = ['Home', 'MyList'];
    const tree = renderer.create(
      <TabBar navigation={navigation} routeNames={routeNames} />,
    );

    expect(tree).toMatchSnapshot();
  });

  it('should assert navigate is called when clicking on a TabBarItem', () => {
    const routeNames = ['Home', 'MyList'];
    const spyNavigationNavigate = jest.spyOn(navigation, 'navigate');
    const tree = renderer.create(
      <TabBar navigation={navigation} routeNames={routeNames} />,
    );

    expect(spyNavigationNavigate).toHaveBeenCalledTimes(0);

    tree.root.findAllByType(TabBarItem)[1].props.onPress('route');

    expect(spyNavigationNavigate).toHaveBeenCalledTimes(1);
    expect(spyNavigationNavigate).toHaveBeenCalledWith('route');
  });
});
