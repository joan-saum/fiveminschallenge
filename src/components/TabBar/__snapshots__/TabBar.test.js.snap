// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBar should render a TabBar 1`] = `
<View
  style={
    Object {
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "space-around",
    }
  }
>
  <View />
  <View />
</View>
`;
